src/Castle.Core/DynamicProxy/Contributors/ClassProxyTargetContributor.cs

```diff
@@ -179,19 +179,10 @@ private Type GetDelegateType(MetaMethod method, ClassEmitter @class, ProxyGenera
 					ToArray(),
 				null);
 
-			var type = scope.GetFromCache(key);
-			if (type != null)
-			{
-				return type;
-			}
-
-			type = new DelegateTypeGenerator(method, targetType)
+			return scope.GetOrAddToCacheWithoutTakingLock(key, _ =>
```

Comment:
I see all these calls to WithoutTakingLock and I wonder - how are you reasoning about whether you hold the lock already or not?

Reply:
I don't. Someone else did that before me, I only refactored pre-existing code.

The general flow of execution appears to be from `IProxyGenerator` into `IProxyBuilder` into "generator" classes which, via their base class `BaseProxyGenerator`, take hold of a lock, then they use contributors to build a model of the proxy type, then they use "emitters" to actually generate code. Most of the calls you're seeing are in the contributors so if all goes well, the lock should be taken at that point.

Reply:
Maybe we should have 'WithoutTakingLock' also do some kind of assertion that the lock is really held by the current thread then? (in debug mode)

Also a different idea - if you wanted the compiler to help analyze whether these assumptions are really correct, you can do an experiment - change 'WithoutTakingLock' to require an extra parameter 'ProofLockHeld', and see what build errors you get. Then fix them (by making callers pass in the parameter, which they must get from their own caller), and repeat - until you find yourself at an ultimate caller where you are taking the lock and acquiring 'ProofLockHeld' by doing so). Might be a painfully slow experiment though! 

And of course it would break all the apis so you can't really ship it this way, or keep it around to prove that nothing is broken ever, so it doesn't have much lasting value ;(

Reply:
> Maybe we should have 'WithoutTakingLock' also do some kind of assertion that the lock is really held by the current thread then?

Perhaps, but I see the present PR purely as a refactoring to get rid of some no-longer-needed types and abstractions. I'd like to keep it focused on that.

Changes helping us resolve #193 don't belong here, in my opinion.

> Also a different idea [...]

It's a good idea as a one-time manual analysis, I might give this a try in a spare moment & separate branch! Thanks. :+1: